crates/server/build.rs

```diff
@@ -1,12 +1,95 @@
-use std::process::Command;
+use std::path::{Path, PathBuf};
+use std::{env, fs};
+
+use cached_path::{Cache, Options};
+use zip::ZipArchive;
 
 fn main() {
-    let status = Command::new("sh")
-        .arg("../../scripts/download-node-ui.sh")
-        .status()
-        .expect("Failed to execute script");
+    let src = option_env!("CALIMERO_WEB_UI_SRC").unwrap_or(
+        "https://github.com/calimero-network/admin-dashboard/archive/refs/heads/master.zip",
+    );
+
+    let force = option_env!("CALIMERO_WEB_UI_FETCH")
+        .map(|v| v == "1" || v.eq_ignore_ascii_case("true"))
+        .unwrap_or(false);
+
+    let project_root = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
+    let static_files_target = project_root.join("../../node-ui/build");
+    let marker_file = project_root.join("build-deps-changed.txt");
+
+    let target_missing = !Path::new("target").exists();
+    let node_ui_missing = !static_files_target.exists();
+
+    if target_missing || node_ui_missing {
+        eprintln!("Triggering rebuild because `target/` or `node-ui/build` is missing.");
+        fs::write(&marker_file, format!("trigger: {:?}\n", std::time::SystemTime::now()))
+            .expect("Failed to write marker file");
+    }
+
+    let cache = Cache::builder()
+        .dir(PathBuf::from(
+            env::var("CARGO_TARGET_DIR").unwrap_or_else(|_| "target".into()),
+        ))
+        .build()
+        .expect("Failed to create cache");
+
+    let options = if force {
+        Options::default().extract()
+    } else {
+        Options::default()
+    };
+
+    let archive_path = cache
+        .cached_path_with_options(src, &options)
+        .expect("Failed to fetch or cache UI archive");
+
+    let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap()).join("web-ui");
+    if out_dir.exists() {
+        fs::remove_dir_all(&out_dir).expect("Failed to remove existing output directory");
+    }
+    fs::create_dir_all(&out_dir).expect("Failed to create output directory");
 
-    if !status.success() {
-        panic!("Script failed with status: {:?}", status);
+    let zip_file = fs::File::open(&archive_path).expect("Cannot open downloaded ZIP archive");
+    let mut zip = ZipArchive::new(zip_file).expect("Failed to read ZIP archive");
+    zip.extract(&out_dir)
+        .expect("Failed to extract ZIP archive");
```

Comment:
what we receive from the `cached_path_with_options` call should be the dir containing the admin dashboard, no further extraction necessary